### PR TITLE
[ClangImporter] Augment type import to support SwiftAttr in type contexts

### LINKED ISSUE #58150: [SR-15885] NS_SWIFT_UI_ACTOR isn't recognized on block typedefs
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15885      |
   |Radar             | None         |
   |Original Reporter | @rjmccall      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 1         |
   |Component/s       |     |
   |Labels            | Bug, StarterBug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: c804c21c1b5258c110b8fb340a9361b7

  </details>

**Issue Description:**

As reported by Aaron Sky on the forums:

<https://forums.swift.org/t/annotate-objective-c-block-type-with-ns-swift-ui-actor/55437>

This method ought to be imported as taking and return a `@MainActor` function type, but doesn't seem to be:

``` java
@import Foundation;

typedef void (^FooAction)(void) NS_SWIFT_UI_ACTOR;

@protocol Foo
@property (nonatomic, nullable) FooAction didDoFoo;
- (void)doFoo;
@end
```

### PR DESCRIPTION
**Pull request by xedin**
clang-side change: https://github.com/apple/llvm-project/pull/7918

Teach `importType` that `clang::AttributedType` now carries an optional `clang::Attr *`
that could be used to express concurrency attributes in type context. 

This PR adds support for the following:

- `__attribute__((swft_attr(...)))` is recognized in nested type contexts i.e. blocks, pointers etc.
- Lightweight generics now support `Sendable` conformance requirement.

There are a couple of refactoring here that make things easier to handle:

- ImportTypeKind::CF{Un}retainedOutParameter became attributes instead.
- New `CompletionHandlerParameter` import kind has been introduced to
  denote completion handler function parameters.
- `SendableCompletionHandlers` is now handled based on type attributes.

Resolves: rdar://119341589
Resolves: rdar://104246144
Fixes #58150

### CODE REVIEW

**Comment by xedin** [comment]
https://github.com/apple/swift/pull/70693
@swift-ci please test

**Comment by xedin** [comment]
https://github.com/apple/llvm-project/pull/7918
@swift-ci please test

**Comment by xedin** [comment]
https://github.com/apple/llvm-project/pull/7918
@swift-ci please test source compatibility

**Review by rjmccall — APPROVED** [review]
Okay, this approach seems reasonable to me.

**Comment by xedin** [comment]
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

**Comment by xedin** [comment]
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

**Comment by xedin** [comment]
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

**Comment by xedin** [comment]
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

**Comment by xedin** [comment]
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

File: lib/ClangImporter/ClangImporter.cpp [hunk 1]
```diff
@@ -7196,36 +7196,40 @@ static bool isForeignReferenceType(const clang::QualType type) {
...
+  return false;
+}
+
 static bool hasOwnedValueAttr(const clang::RecordDecl *decl) {
-  return decl->hasAttrs() && llvm::any_of(decl->getAttrs(), [](auto *attr) {
-           if (auto swiftAttr = dyn_cast<clang::SwiftAttrAttr>(attr))
-             return swiftAttr->getAttribute() == "import_owned";
-           return false;
-         });
+  return hasSwiftAttribute(decl, "import_owned");
```
Comment (xedin) [inline lib/ClangImporter/ClangImporter.cpp:7220]: cc @egorzhdan

**Pull request merged**